pkg/api/v1/handler/oauth.go

```diff
@@ -0,0 +1,111 @@
+package handler
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/caicloud/cyclone/cmd/worker/options"
+	"github.com/caicloud/cyclone/pkg/api"
+	"github.com/caicloud/cyclone/pkg/util/os"
+	"github.com/caicloud/nirvana/log"
+	"github.com/caicloud/nirvana/service"
+	"golang.org/x/oauth2"
+)
+
+const ()
+
+var gitlabConf = &oauth2.Config{
+	ClientID:     os.GetStringEnv(options.GitlabClient, ""),
+	ClientSecret: os.GetStringEnv(options.GitlabSecret, ""),
+	RedirectURL:  os.GetStringEnv(options.CycloneServer, "") + "/api/v1/scms/gitlab/callback",
+	Scopes:       []string{},
+	Endpoint: oauth2.Endpoint{
+		AuthURL:  fmt.Sprintf("%s/oauth/authorize", os.GetStringEnv(options.GitlabURL, "")),
+		TokenURL: fmt.Sprintf("%s/oauth/token", os.GetStringEnv(options.GitlabURL, "")),
+	},
+}
+
+var fakeState = "fakestate"
+
+// GetAuthCodeURL get the redirect url from config
+// then return url to frontend
+func GetAuthCodeURL(ctx context.Context, scm string) (string, error) {
+	switch scm {
+	case "gitlab":
+		url := gitlabConf.AuthCodeURL(fakeState)
+		log.Infof("new redirect url for gitlab oauth2: %s", url)
+		return url, nil
+	case "github":
+		//TODO: support github oauth2
+		return "", nil
```

Comment:
I think to to return Not Implemented error is better than nil for github, and add default case is needed.